File: src/primitives/Hash/keccak256.js
import { keccak_256 } from "@noble/hashes/sha3.js";

/**
 * Hash data with Keccak-256
 *
 * @param {Uint8Array} data - Data to hash
 * @returns {import('./BrandedHash.js').BrandedHash} 32-byte hash
 *
 * @example
 * ```js
 * const hash = Hash.keccak256(data);
 * ```
 */
export function keccak256(data) {
	return /** @type {import('./BrandedHash.ts').BrandedHash} */ (
		keccak_256(data)
	);
}

File: packages/voltaire-go/crypto/keccak256/keccak256.go
// Package keccak256 provides Keccak-256 hashing.
package keccak256

import (
	"github.com/voltaire-labs/voltaire-go/internal/ffi"
	"github.com/voltaire-labs/voltaire-go/primitives/hash"
)

// Hash computes the Keccak-256 hash of data.
func Hash(data []byte) hash.Hash {
	return hash.Hash(ffi.Keccak256(data))
}

// HashString computes the Keccak-256 hash of a UTF-8 string.
func HashString(s string) hash.Hash {
	return Hash([]byte(s))
}

// Sum returns the Keccak-256 hash of multiple byte slices concatenated.
func Sum(data ...[]byte) hash.Hash {
	total := 0
	for _, d := range data {
		total += len(d)
	}
	combined := make([]byte, 0, total)
	for _, d := range data {
		combined = append(combined, d...)
	}
	return Hash(combined)
}
